packages/itwinui-react/src/core/utils/components/ShadowRoot.tsx

```diff
@@ -85,27 +86,34 @@ function useShadowRoot(
       return;
     }
 
-    if (parent.shadowRoot) {
-      parent.shadowRoot.replaceChildren(); // Remove previous shadowroot content
-    }
+    const setupShadowRoot = () => {
+      if (parent.shadowRoot) {
+        parent.shadowRoot.replaceChildren(); // Remove previous shadowroot content
+      }
 
-    const shadow = parent.shadowRoot || parent.attachShadow({ mode: 'open' });
+      const shadow = parent.shadowRoot || parent.attachShadow({ mode: 'open' });
 
-    if (supportsAdoptedStylesheets) {
-      // create an empty stylesheet and add it to the shadowRoot
-      const currentWindow = shadow.ownerDocument.defaultView || globalThis;
-      styleSheet.current = new currentWindow.CSSStyleSheet();
-      shadow.adoptedStyleSheets = [styleSheet.current];
+      if (supportsAdoptedStylesheets) {
+        // create an empty stylesheet and add it to the shadowRoot
+        const currentWindow = shadow.ownerDocument.defaultView || globalThis;
+        styleSheet.current = new currentWindow.CSSStyleSheet();
+        shadow.adoptedStyleSheets = [styleSheet.current];
 
-      // add the CSS immediately to avoid FOUC (one-time)
-      if (latestCss.current) {
-        styleSheet.current.replaceSync(latestCss.current);
+        // add the CSS immediately to avoid FOUC (one-time)
+        if (latestCss.current) {
+          styleSheet.current.replaceSync(latestCss.current);
+        }
       }
-    }
 
-    queueMicrotask(() => {
-      // Flush the state immediately to ensure layout measurements in parent component are correct
+      // Flush the state immediately after shadow-root is attached, to ensure that layout
+      // measurements in parent component are correct.
+      // Without this, the parent component may end up measuring the layout when the shadow-root
+      // is attached in the DOM but React hasn't rendered any slots or content into it yet.
       ReactDOM.flushSync(() => setShadowRoot(shadow));
+    };
+
+    queueMicrotask(() => {
+      setupShadowRoot();
```

Comment:
By fixing the first client render, I was able to reproduce the original issue related to incorrect measurements in Dialog. It was trying to read measurements when the shadow-root was attached but state wasn't updated.

![image](https://github.com/iTwin/iTwinUI/assets/9084735/07921a0c-36d8-482e-ab42-5df82f0b8cfa)

To fix it, I moved both of these things (`attachShadow` browser call and `setShadowRoot` react call) closer to each other into queueMicrotask. (Same as it was before #1954)